src/input/devices.rs

```diff
@@ -0,0 +1,411 @@
+use std::{
+    ffi::CStr,
+    sync::{
+        atomic::{AtomicBool, Ordering},
+        Mutex,
+    },
+};
+
+use openvr as vr;
+use openxr as xr;
+
+use crate::openxr_data::{self, AtomicPath, Hand, OpenXrData, SessionData};
+use crate::tracy_span;
+use log::trace;
+
+use super::{profiles::MainAxisType, Input, InteractionProfile};
+
+#[derive(Debug, Copy, Clone, PartialEq)]
+pub enum TrackedDeviceType {
+    Hmd,
+    Controller { hand: Hand },
+}
+pub struct XrTrackedDevice {
+    device_type: TrackedDeviceType,
+    interaction_profile: Mutex<Option<&'static dyn InteractionProfile>>,
+    profile_path: AtomicPath,
+    connected: AtomicBool,
+    previous_connected: AtomicBool,
+    pose_cache: Mutex<Option<vr::TrackedDevicePose_t>>,
+}
+
+pub struct TrackedDeviceCreateInfo {
+    pub device_type: TrackedDeviceType,
+    pub profile_path: Option<xr::Path>,
+    pub interaction_profile: Option<&'static dyn InteractionProfile>,
+}
+
+impl XrTrackedDevice {
+    pub fn new(info: TrackedDeviceCreateInfo) -> Self {
+        let profile_path = AtomicPath::new();
+
+        if let Some(path) = info.profile_path {
+            profile_path.store(path);
+        }
+
+        Self {
+            device_type: info.device_type,
+            interaction_profile: Mutex::new(info.interaction_profile),
+            profile_path,
+            connected: if info.device_type == TrackedDeviceType::Hmd {
+                true.into()
+            } else {
+                false.into()
+            },
+            previous_connected: false.into(),
+            pose_cache: Mutex::new(None),
+        }
+    }
+
+    pub fn get_pose(
+        &self,
+        xr_data: &OpenXrData<impl crate::openxr_data::Compositor>,
+        session_data: &SessionData,
+        origin: vr::ETrackingUniverseOrigin,
+    ) -> Option<vr::TrackedDevicePose_t> {
+        let mut pose_cache = self.pose_cache.lock().ok()?;
```

Comment:
```suggestion
        let mut pose_cache = self.pose_cache.lock().unwrap();
```
Mutex poisoning is not something I intend to ever try and recover from.